design/design-doc-composition-functions.md

```diff
@@ -0,0 +1,599 @@
+gVisor be used.
+
+Testing using EKS v1.21.5-eks-9017834 with Amazon Linux 2 nodes (Kernel
+5.4.188-104.359.amzn2.x86_64) found that it was possible to run `unshare -rUm`
+inside an Alpine Linux container 'out of the box'.
+
+The `unshare` syscall used to create containers is rejected by the default
+Docker and containerd seccomp profiles. seccomp is disabled ("Unconstrained") by
+default in Kubernetes, but that will soon change per [this KEP][kep-seccomp]
+which proposes that Kubernetes use the seccomp profiles of its container engine
+(i.e. containerd) by default. Once this happens Crossplane will either need to
+run with the "Unconstrained" seccomp profile, or a variant of the default
+containerd seccomp profile that allows a few extra syscalls (i.e. at least
+`unshare` and `mount`). This can be done by setting a Pod's
+`spec.securityContext.seccompProfile.type` field to `Unconstrained`.
+
+### Packaging Containerized Functions
+
+This document proposes that containerized functions support Crossplane [package
+metadata][package-meta] in the form of a `package.yaml` file at the root of the
+flattened filesystem and/or the OCI layer annotated as `io.crossplane.xpkg:
+base` per the [xpkg spec][xpkg-spec]. This `package.yaml` file would contain a
+custom-resource-like YAML document of type `Function.meta.pkg.crossplane.io`.
+
+Unlike `Configuration` and `Provider` packages, `Function` packages would not
+actually be processed by the Crossplane package manager but rather by the
+Composition (`apiextensions`) machinery. In practice Crossplane would be
+ignorant of the `package.yaml` file; it would exist purely as a way to attach
+"package-like" metadata to containerized Crossplane functions. Therefore unlike
+the existing package types the `package.yaml` would contain no `spec` section.
+
+An example `package.yaml` might look like:
+
+```yaml
+# Required. Must be as below.
+apiVersion: meta.pkg.crossplane.io/v1alpha1
+# Required. Must be as below.
+kind: Function
+# Required.
+metadata:
+  # Required. Must comply with Kubernetes API conventions.
+  name: function-example
+  # Optional. Must comply with Kubernetes API conventions.
+  annotations:
+    meta.crossplane.io/source: https://github.com/negz/example-fn
+    meta.crossplane.io/description: An example function
+```
+
+## Alternatives Considered
+
+Most of the alternatives considered in this design could also be thought of as
+future considerations. In most cases these alternatives don't make sense at the
+present time but likely will in future.
+
+### Using Webhooks to Run Functions
+
+Crossplane could invoke functions by calling a webhook rather than running an
+OCI container. In this model function input and output would still take the form
+of a `ResourceList`, but would be HTTP request and response bodies rather than a
+container's stdin and stdout.
+
+The primary detractor of this approach is the burden it puts on function authors
```

Comment:
While I see this as a valid concern, the complexities of running and owning the lifecycle of a whole function approach in Kubernetes inside Crossplane sounds like an overkill. I am of course biased because I am working on Knative Functions, but if I have a clear interface from Crossplane with inputs and outputs, the functions side of things is already optimized by other projects, so running and operating these functions is already the concern of these other frameworks. The approach that the project MetaController took follows this webhook approach and it took me 10 minutes to create a function to create a metacontroller controller. Notice the big overlap of APIs between metacontroller (https://metacontroller.github.io/metacontroller/guide/create.html) and this proposal, because I do see that implementing what this proposal is describing around running functions and pipelines of functions seems extremely hard to implement and messy. I will be happy to have a call about this in more detail if needed.

Reply:
> Running and owning the lifecycle of a whole function approach [...] inside Crossplane

FWIW we'll likely iterate on this design to reflect the function runner being a swappable sidecar with Crossplane speaking e.g. gRPC to invoke functions, rather than something that literally happens inside the Crossplane process. I do still feel strongly that said sidecar should be invoking OCI images though. We also considered having a webhook that would run OCI based functions but the `Composition` API is weaker in this situation when Crossplane itself can't speak in OCI image terms.